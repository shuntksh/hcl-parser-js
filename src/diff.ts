/**
 * Helper function to walk through two objects and report their differences
 * It is used to compare two HCL ASTs as they tend to be very similar but not identical
 * @param obj1 First object to compare
 * @param obj2 Second object to compare
 * @param path Current path in the object (used for recursion)
 * @returns Array of differences with their paths
 */
export function diff(obj1: unknown, obj2: unknown, path = ""): string[] {
	// If both objects are identical, return empty array
	if (obj1 === obj2) return [];

	// If either is null/undefined but not both
	if (
		obj1 === null ||
		obj2 === null ||
		obj1 === undefined ||
		obj2 === undefined
	) {
		return [`${path}: ${obj1} !== ${obj2}`];
	}

	// If types don't match
	if (typeof obj1 !== typeof obj2) {
		return [`${path}: type mismatch - ${typeof obj1} !== ${typeof obj2}`];
	}

	// If not objects, compare values
	if (typeof obj1 !== "object" || typeof obj2 !== "object") {
		return obj1 === obj2 ? [] : [`${path}: ${obj1} !== ${obj2}`];
	}

	// Handle arrays
	if (Array.isArray(obj1) && Array.isArray(obj2)) {
		if (obj1.length !== obj2.length) {
			return [
				`${path}: array length mismatch - ${obj1.length} !== ${obj2.length}`,
			];
		}

		const diffs: string[] = [];
		for (let i = 0; i < obj1.length; i++) {
			const nestedDiffs = diff(obj1[i], obj2[i], `${path}[${i}]`);
			diffs.push(...nestedDiffs);
		}
		return diffs;
	}

	// Handle regular objects
	if (!Array.isArray(obj1) && !Array.isArray(obj2)) {
		const diffs: string[] = [];
		const allKeys = new Set([
			...Object.keys(obj1 as object),
			...Object.keys(obj2 as object),
		]);

		for (const key of allKeys) {
			const obj1Value = (obj1 as Record<string, unknown>)[key];
			const obj2Value = (obj2 as Record<string, unknown>)[key];

			// Key exists in both objects
			if (key in (obj1 as object) && key in (obj2 as object)) {
				const nestedDiffs = diff(
					obj1Value,
					obj2Value,
					path ? `${path}.${key}` : key,
				);
				diffs.push(...nestedDiffs);
			}
			// Key only exists in obj1
			else if (key in (obj1 as object)) {
				diffs.push(
					`${path ? `${path}.${key}` : key}: missing in second object`,
				);
			}
			// Key only exists in obj2
			else {
				diffs.push(`${path ? `${path}.${key}` : key}: missing in first object`);
			}
		}
		return diffs;
	}

	// If we get here, one is an array and one is an object
	return [`${path}: structure mismatch - array vs object`];
}
