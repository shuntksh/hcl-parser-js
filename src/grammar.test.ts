import { describe, expect, test } from "bun:test";
import path from "node:path";

import { parse } from "./__generated__";
import { diff } from "./diff";
import { schema } from "./schema";
import { stringify } from "./stringify";

describe("Smoke", () => {
	test("parses a test tf file", async () => {
		const template = await Bun.file(
			path.join(__dirname, "./grammar.test.hcl"),
		).text();
		const ast = parse(template);
		expect(ast).toBeObject();
		expect(ast).toMatchSnapshot();
		expect(schema.parse(ast)).toEqual(ast);
	});

	test("throws on invalid syntax", () => {
		const input = `resource aws_instance" "web" {}`; // Missing opening quote
		expect(() => parse(input)).toThrow();
	});
});
