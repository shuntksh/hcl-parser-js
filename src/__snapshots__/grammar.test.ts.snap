// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Smoke parses a test tf file 1`] = `
[
  {
    "attribute": {
      "name": {
        "type": "Identifier",
        "value": "type",
      },
      "type": "Attribute",
      "value": {
        "name": {
          "type": "Identifier",
          "value": "bool",
        },
        "type": "VariableExpression",
      },
    },
    "blockType": {
      "type": "Identifier",
      "value": "variable",
    },
    "labels": [
      {
        "type": "StringLiteral",
        "value": "online",
      },
    ],
    "type": "OneLineBlock",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "variable",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "type",
        },
        "type": "Attribute",
        "value": {
          "name": {
            "type": "Identifier",
            "value": "number",
          },
          "type": "VariableExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "default",
        },
        "type": "Attribute",
        "value": {
          "type": "NumberLiteral",
          "value": 8080,
        },
      },
    ],
    "labels": [
      {
        "type": "StringLiteral",
        "value": "number_var",
      },
    ],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "variable",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "type",
        },
        "type": "Attribute",
        "value": {
          "args": [
            {
              "name": {
                "type": "Identifier",
                "value": "string",
              },
              "type": "VariableExpression",
            },
          ],
          "name": {
            "type": "Identifier",
            "value": "list",
          },
          "type": "FunctionCallExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "default",
        },
        "type": "Attribute",
        "value": {
          "elements": [
            {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "item1",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
            {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "item2",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
            {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "item3",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
          ],
          "type": "TupleValue",
        },
      },
    ],
    "labels": [
      {
        "type": "StringLiteral",
        "value": "tuple_var",
      },
    ],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "variable",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "type",
        },
        "type": "Attribute",
        "value": {
          "name": {
            "type": "Identifier",
            "value": "string",
          },
          "type": "VariableExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "default",
        },
        "type": "Attribute",
        "value": {
          "parts": [
            {
              "type": "TemplateLiteral",
              "value": "vpc-123456",
            },
          ],
          "type": "QuotedTemplateExpression",
        },
      },
    ],
    "labels": [
      {
        "type": "StringLiteral",
        "value": "string_var",
      },
    ],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "variable",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "type",
        },
        "type": "Attribute",
        "value": {
          "args": [
            {
              "name": {
                "type": "Identifier",
                "value": "string",
              },
              "type": "VariableExpression",
            },
          ],
          "name": {
            "type": "Identifier",
            "value": "map",
          },
          "type": "FunctionCallExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "default",
        },
        "type": "Attribute",
        "value": {
          "elements": [
            {
              "key": {
                "type": "Identifier",
                "value": "Environment",
              },
              "value": {
                "parts": [
                  {
                    "type": "TemplateLiteral",
                    "value": "test",
                  },
                ],
                "type": "QuotedTemplateExpression",
              },
            },
            {
              "key": {
                "type": "Identifier",
                "value": "Managed_by",
              },
              "value": {
                "parts": [
                  {
                    "type": "TemplateLiteral",
                    "value": "terraform",
                  },
                ],
                "type": "QuotedTemplateExpression",
              },
            },
          ],
          "type": "ObjectValue",
        },
      },
    ],
    "labels": [
      {
        "type": "StringLiteral",
        "value": "dictionary_var",
      },
    ],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "data",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "string_val",
        },
        "type": "Attribute",
        "value": {
          "parts": [
            {
              "type": "TemplateLiteral",
              "value": "simple string",
            },
          ],
          "type": "QuotedTemplateExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "number_val",
        },
        "type": "Attribute",
        "value": {
          "type": "NumberLiteral",
          "value": 42,
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "bool_val",
        },
        "type": "Attribute",
        "value": {
          "type": "BooleanLiteral",
          "value": true,
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "null_val",
        },
        "type": "Attribute",
        "value": {
          "type": "NullLiteral",
          "value": null,
        },
      },
    ],
    "labels": [
      {
        "type": "StringLiteral",
        "value": "data",
      },
      {
        "type": "StringLiteral",
        "value": "basic_types",
      },
    ],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "data",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "interpolated",
        },
        "type": "Attribute",
        "value": {
          "parts": [
            {
              "type": "TemplateLiteral",
              "value": "Hello, ",
            },
            {
              "expression": {
                "key": {
                  "type": "Identifier",
                  "value": "name",
                },
                "target": {
                  "name": {
                    "type": "Identifier",
                    "value": "var",
                  },
                  "type": "VariableExpression",
                },
                "type": "GetAttributeOperator",
              },
              "strip": {
                "left": false,
                "right": false,
              },
              "type": "TemplateInterpolation",
            },
            {
              "type": "TemplateLiteral",
              "value": "!",
            },
          ],
          "type": "QuotedTemplateExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "multiline",
        },
        "type": "Attribute",
        "value": {
          "marker": {
            "type": "Identifier",
            "value": "EOT",
          },
          "stripIndent": true,
          "template": [
            {
              "type": "TemplateLiteral",
              "value": "    Server: ",
            },
            {
              "expression": {
                "key": {
                  "type": "Identifier",
                  "value": "hostname",
                },
                "target": {
                  "name": {
                    "type": "Identifier",
                    "value": "var",
                  },
                  "type": "VariableExpression",
                },
                "type": "GetAttributeOperator",
              },
              "strip": {
                "left": false,
                "right": false,
              },
              "type": "TemplateInterpolation",
            },
            {
              "type": "TemplateLiteral",
              "value": 
"
    Port: "
,
            },
            {
              "expression": {
                "key": {
                  "type": "Identifier",
                  "value": "port",
                },
                "target": {
                  "name": {
                    "type": "Identifier",
                    "value": "var",
                  },
                  "type": "VariableExpression",
                },
                "type": "GetAttributeOperator",
              },
              "strip": {
                "left": false,
                "right": false,
              },
              "type": "TemplateInterpolation",
            },
            {
              "type": "TemplateLiteral",
              "value": 
"
    Active: "
,
            },
            {
              "expression": {
                "type": "BooleanLiteral",
                "value": true,
              },
              "strip": {
                "left": false,
                "right": false,
              },
              "type": "TemplateInterpolation",
            },
          ],
          "type": "HeredocTemplateExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "raw_heredoc",
        },
        "type": "Attribute",
        "value": {
          "marker": {
            "type": "Identifier",
            "value": "EOT",
          },
          "stripIndent": false,
          "template": [
            {
              "type": "TemplateLiteral",
              "value": 
"    Keep this indentation
      exactly as is"
,
            },
          ],
          "type": "HeredocTemplateExpression",
        },
      },
    ],
    "labels": [
      {
        "type": "StringLiteral",
        "value": "data",
      },
      {
        "type": "StringLiteral",
        "value": "string_expressions",
      },
    ],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "data",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "conditional",
        },
        "type": "Attribute",
        "value": {
          "left": {
            "key": {
              "type": "Identifier",
              "value": "environment",
            },
            "target": {
              "name": {
                "type": "Identifier",
                "value": "var",
              },
              "type": "VariableExpression",
            },
            "type": "GetAttributeOperator",
          },
          "operator": "==",
          "right": {
            "falseExpr": {
              "type": "NumberLiteral",
              "value": 1,
            },
            "predicate": {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "prod",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
            "trueExpr": {
              "type": "NumberLiteral",
              "value": 5,
            },
            "type": "ConditionalOperator",
          },
          "type": "BinaryOperator",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "arithmetic",
        },
        "type": "Attribute",
        "value": {
          "left": {
            "left": {
              "type": "NumberLiteral",
              "value": 10,
            },
            "operator": "*",
            "right": {
              "type": "NumberLiteral",
              "value": 5,
            },
            "type": "BinaryOperator",
          },
          "operator": "+",
          "right": {
            "left": {
              "type": "NumberLiteral",
              "value": 20,
            },
            "operator": "/",
            "right": {
              "type": "NumberLiteral",
              "value": 2,
            },
            "type": "BinaryOperator",
          },
          "type": "BinaryOperator",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "comparison",
        },
        "type": "Attribute",
        "value": {
          "left": {
            "type": "NumberLiteral",
            "value": 5,
          },
          "operator": ">",
          "right": {
            "left": {
              "type": "NumberLiteral",
              "value": 3,
            },
            "operator": "&&",
            "right": {
              "left": {
                "type": "NumberLiteral",
                "value": 2,
              },
              "operator": "<=",
              "right": {
                "type": "NumberLiteral",
                "value": 4,
              },
              "type": "BinaryOperator",
            },
            "type": "BinaryOperator",
          },
          "type": "BinaryOperator",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "negation",
        },
        "type": "Attribute",
        "value": {
          "operator": "!",
          "term": {
            "type": "BooleanLiteral",
            "value": false,
          },
          "type": "UnaryOperator",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "negative",
        },
        "type": "Attribute",
        "value": {
          "operator": "-",
          "term": {
            "type": "NumberLiteral",
            "value": 42,
          },
          "type": "UnaryOperator",
        },
      },
    ],
    "labels": [
      {
        "type": "StringLiteral",
        "value": "data",
      },
      {
        "type": "StringLiteral",
        "value": "expressions",
      },
    ],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "data",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "simple_list",
        },
        "type": "Attribute",
        "value": {
          "elements": [
            {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "a",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
            {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "b",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
            {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "c",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
          ],
          "type": "TupleValue",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "mixed_tuple",
        },
        "type": "Attribute",
        "value": {
          "elements": [
            {
              "type": "NumberLiteral",
              "value": 1,
            },
            {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "two",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
            {
              "type": "BooleanLiteral",
              "value": true,
            },
          ],
          "type": "TupleValue",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "simple_map",
        },
        "type": "Attribute",
        "value": {
          "elements": [
            {
              "key": {
                "type": "Identifier",
                "value": "key1",
              },
              "value": {
                "parts": [
                  {
                    "type": "TemplateLiteral",
                    "value": "value1",
                  },
                ],
                "type": "QuotedTemplateExpression",
              },
            },
            {
              "key": {
                "type": "Identifier",
                "value": "key2",
              },
              "value": {
                "parts": [
                  {
                    "type": "TemplateLiteral",
                    "value": "value2",
                  },
                ],
                "type": "QuotedTemplateExpression",
              },
            },
          ],
          "type": "ObjectValue",
        },
      },
    ],
    "labels": [
      {
        "type": "StringLiteral",
        "value": "data",
      },
      {
        "type": "StringLiteral",
        "value": "collections",
      },
    ],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "locals",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "tuple_transform",
        },
        "type": "Attribute",
        "value": {
          "condition": {
            "left": {
              "name": {
                "type": "Identifier",
                "value": "item",
              },
              "type": "VariableExpression",
            },
            "operator": "!=",
            "right": {
              "parts": [],
              "type": "QuotedTemplateExpression",
            },
            "type": "BinaryOperator",
          },
          "expression": {
            "args": [
              {
                "name": {
                  "type": "Identifier",
                  "value": "item",
                },
                "type": "VariableExpression",
              },
            ],
            "name": {
              "type": "Identifier",
              "value": "upper",
            },
            "type": "FunctionCallExpression",
          },
          "intro": {
            "collection": {
              "key": {
                "type": "Identifier",
                "value": "tuple_var",
              },
              "target": {
                "name": {
                  "type": "Identifier",
                  "value": "var",
                },
                "type": "VariableExpression",
              },
              "type": "GetAttributeOperator",
            },
            "iterator": {
              "type": "Identifier",
              "value": "item",
            },
            "value": null,
          },
          "kind": "tuple",
          "type": "ForExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "object_transform",
        },
        "type": "Attribute",
        "value": {
          "condition": {
            "left": {
              "name": {
                "type": "Identifier",
                "value": "v",
              },
              "type": "VariableExpression",
            },
            "operator": "!=",
            "right": {
              "type": "NullLiteral",
              "value": null,
            },
            "type": "BinaryOperator",
          },
          "grouping": false,
          "intro": {
            "collection": {
              "key": {
                "type": "Identifier",
                "value": "dictionary_var",
              },
              "target": {
                "name": {
                  "type": "Identifier",
                  "value": "var",
                },
                "type": "VariableExpression",
              },
              "type": "GetAttributeOperator",
            },
            "iterator": {
              "type": "Identifier",
              "value": "k",
            },
            "value": null,
          },
          "key": {
            "name": {
              "type": "Identifier",
              "value": "k",
            },
            "type": "VariableExpression",
          },
          "kind": "object",
          "type": "ForExpression",
          "value": {
            "args": [
              {
                "name": {
                  "type": "Identifier",
                  "value": "v",
                },
                "type": "VariableExpression",
              },
            ],
            "name": {
              "type": "Identifier",
              "value": "upper",
            },
            "type": "FunctionCallExpression",
          },
        },
      },
    ],
    "labels": [],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "resource",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "name",
        },
        "type": "Attribute",
        "value": {
          "parts": [
            {
              "type": "TemplateLiteral",
              "value": "test-sg",
            },
          ],
          "type": "QuotedTemplateExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "description",
        },
        "type": "Attribute",
        "value": {
          "parts": [
            {
              "type": "TemplateLiteral",
              "value": "Test security group",
            },
          ],
          "type": "QuotedTemplateExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "vpc_id",
        },
        "type": "Attribute",
        "value": {
          "key": {
            "type": "Identifier",
            "value": "string_var",
          },
          "target": {
            "name": {
              "type": "Identifier",
              "value": "var",
            },
            "type": "VariableExpression",
          },
          "type": "GetAttributeOperator",
        },
      },
      {
        "blockType": {
          "type": "Identifier",
          "value": "ingress",
        },
        "bodies": [
          {
            "name": {
              "type": "Identifier",
              "value": "from_port",
            },
            "type": "Attribute",
            "value": {
              "type": "NumberLiteral",
              "value": 80,
            },
          },
          {
            "name": {
              "type": "Identifier",
              "value": "to_port",
            },
            "type": "Attribute",
            "value": {
              "type": "NumberLiteral",
              "value": 80,
            },
          },
          {
            "name": {
              "type": "Identifier",
              "value": "protocol",
            },
            "type": "Attribute",
            "value": {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "tcp",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
          },
        ],
        "labels": [],
        "type": "Block",
      },
      {
        "blockType": {
          "type": "Identifier",
          "value": "ingress",
        },
        "bodies": [
          {
            "name": {
              "type": "Identifier",
              "value": "from_port",
            },
            "type": "Attribute",
            "value": {
              "type": "NumberLiteral",
              "value": 443,
            },
          },
          {
            "name": {
              "type": "Identifier",
              "value": "to_port",
            },
            "type": "Attribute",
            "value": {
              "type": "NumberLiteral",
              "value": 443,
            },
          },
          {
            "name": {
              "type": "Identifier",
              "value": "protocol",
            },
            "type": "Attribute",
            "value": {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "tcp",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
          },
        ],
        "labels": [],
        "type": "Block",
      },
    ],
    "labels": [
      {
        "type": "StringLiteral",
        "value": "aws_security_group",
      },
      {
        "type": "StringLiteral",
        "value": "test",
      },
    ],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "resource",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "name",
        },
        "type": "Attribute",
        "value": {
          "parts": [
            {
              "type": "TemplateLiteral",
              "value": "test-asg",
            },
          ],
          "type": "QuotedTemplateExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "min_size",
        },
        "type": "Attribute",
        "value": {
          "type": "NumberLiteral",
          "value": 1,
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "max_size",
        },
        "type": "Attribute",
        "value": {
          "type": "NumberLiteral",
          "value": 10,
        },
      },
      {
        "blockType": {
          "type": "Identifier",
          "value": "dynamic",
        },
        "bodies": [
          {
            "name": {
              "type": "Identifier",
              "value": "for_each",
            },
            "type": "Attribute",
            "value": {
              "key": {
                "type": "Identifier",
                "value": "dictionary_var",
              },
              "target": {
                "name": {
                  "type": "Identifier",
                  "value": "var",
                },
                "type": "VariableExpression",
              },
              "type": "GetAttributeOperator",
            },
          },
          {
            "blockType": {
              "type": "Identifier",
              "value": "content",
            },
            "bodies": [
              {
                "name": {
                  "type": "Identifier",
                  "value": "key",
                },
                "type": "Attribute",
                "value": {
                  "key": {
                    "type": "Identifier",
                    "value": "key",
                  },
                  "target": {
                    "name": {
                      "type": "Identifier",
                      "value": "tag",
                    },
                    "type": "VariableExpression",
                  },
                  "type": "GetAttributeOperator",
                },
              },
              {
                "name": {
                  "type": "Identifier",
                  "value": "value",
                },
                "type": "Attribute",
                "value": {
                  "key": {
                    "type": "Identifier",
                    "value": "value",
                  },
                  "target": {
                    "name": {
                      "type": "Identifier",
                      "value": "tag",
                    },
                    "type": "VariableExpression",
                  },
                  "type": "GetAttributeOperator",
                },
              },
              {
                "name": {
                  "type": "Identifier",
                  "value": "propagate_at_launch",
                },
                "type": "Attribute",
                "value": {
                  "type": "BooleanLiteral",
                  "value": true,
                },
              },
            ],
            "labels": [],
            "type": "Block",
          },
        ],
        "labels": [
          {
            "type": "StringLiteral",
            "value": "tag",
          },
        ],
        "type": "Block",
      },
    ],
    "labels": [
      {
        "type": "StringLiteral",
        "value": "aws_autoscaling_group",
      },
      {
        "type": "StringLiteral",
        "value": "test",
      },
    ],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "locals",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "timestamp",
        },
        "type": "Attribute",
        "value": {
          "args": [],
          "name": {
            "type": "Identifier",
            "value": "timestamp",
          },
          "type": "FunctionCallExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "upper",
        },
        "type": "Attribute",
        "value": {
          "args": [
            {
              "parts": [
                {
                  "type": "TemplateLiteral",
                  "value": "test",
                },
              ],
              "type": "QuotedTemplateExpression",
            },
          ],
          "name": {
            "type": "Identifier",
            "value": "upper",
          },
          "type": "FunctionCallExpression",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "encoded",
        },
        "type": "Attribute",
        "value": {
          "args": [
            {
              "args": [
                {
                  "elements": [
                    {
                      "key": {
                        "type": "Identifier",
                        "value": "foo",
                      },
                      "value": {
                        "parts": [
                          {
                            "type": "TemplateLiteral",
                            "value": "bar",
                          },
                        ],
                        "type": "QuotedTemplateExpression",
                      },
                    },
                    {
                      "key": {
                        "type": "Identifier",
                        "value": "baz",
                      },
                      "value": {
                        "type": "NumberLiteral",
                        "value": 123,
                      },
                    },
                  ],
                  "type": "ObjectValue",
                },
              ],
              "name": {
                "type": "Identifier",
                "value": "jsonencode",
              },
              "type": "FunctionCallExpression",
            },
          ],
          "name": {
            "type": "Identifier",
            "value": "base64encode",
          },
          "type": "FunctionCallExpression",
        },
      },
    ],
    "labels": [],
    "type": "Block",
  },
  {
    "blockType": {
      "type": "Identifier",
      "value": "locals",
    },
    "bodies": [
      {
        "name": {
          "type": "Identifier",
          "value": "instance_ids",
        },
        "type": "Attribute",
        "value": {
          "attributes": [
            {
              "key": {
                "type": "Identifier",
                "value": "id",
              },
              "type": "GetAttributeOperator",
            },
          ],
          "kind": "Attribute",
          "target": {
            "key": {
              "type": "Identifier",
              "value": "test",
            },
            "target": {
              "name": {
                "type": "Identifier",
                "value": "aws_instance",
              },
              "type": "VariableExpression",
            },
            "type": "GetAttributeOperator",
          },
          "type": "SplatOperator",
        },
      },
      {
        "name": {
          "type": "Identifier",
          "value": "public_ips",
        },
        "type": "Attribute",
        "value": {
          "condition": null,
          "expression": {
            "key": {
              "type": "Identifier",
              "value": "public_ip",
            },
            "target": {
              "name": {
                "type": "Identifier",
                "value": "i",
              },
              "type": "VariableExpression",
            },
            "type": "GetAttributeOperator",
          },
          "intro": {
            "collection": {
              "key": {
                "type": "Identifier",
                "value": "test",
              },
              "target": {
                "name": {
                  "type": "Identifier",
                  "value": "aws_instance",
                },
                "type": "VariableExpression",
              },
              "type": "GetAttributeOperator",
            },
            "iterator": {
              "type": "Identifier",
              "value": "i",
            },
            "value": null,
          },
          "kind": "tuple",
          "type": "ForExpression",
        },
      },
    ],
    "labels": [],
    "type": "Block",
  },
]
`;
